### ISSUE TITLE
accessor_iterator_requirement.cpp uses virtual functions on device side

### ISSUE THREAD

**Issue by AlexeySachkov** [KhronosGroup/SYCL-CTS#388]
Hi,

We recently added accessor iterators support to intel/llvm repo and I've been trying to enable corresponding CTS tests. Unfortunately, they can't be compiled due to some errors in them and while some of them are trivial to fix, there is one I would like to discuss first.

As the title says, there is a code within those tests, which uses virtual functions on device side of the test. It originates from `WARN` macro we have in some tests to indicate that coverage might not be full due to parameters passed to a helper function.

Error example:
```
In file included from KhronosGroup/SYCL-CTS/tests/accessor/accessor_iterator_requirement.cpp:9:
In file included from KhronosGroup/SYCL-CTS/tests/accessor/../../util/named_requirement_verification/legacy
_random_access_iterator.h:14:
In file included from KhronosGroup/SYCL-CTS/tests/accessor/../../util/named_requirement_verification/legacy
_bidirectional_iterator.h:14:
In file included from KhronosGroup/SYCL-CTS/tests/accessor/../../util/named_requirement_verification/legacy
_forward_iterator.h:17:
In file included from KhronosGroup/SYCL-CTS/vendor/Catch2/src/catch2/../catch2/catch_test_macros.hpp:11:
In file included from KhronosGroup/SYCL-CTS/vendor/Catch2/src/catch2/../catch2/internal/catch_test_macro_im
pl.hpp:11:
KhronosGroup/SYCL-CTS/vendor/Catch2/src/catch2/../catch2/internal/catch_assertion_handler.hpp:41:33: error:
·
      SYCL kernel cannot call a virtual function
                m_resultCapture.handleIncomplete( m_assertionInfo );
                                ^
In file included from KhronosGroup/SYCL-CTS/tests/accessor/accessor_iterator_requirement.cpp:9:
In file included from KhronosGroup/SYCL-CTS/tests/accessor/../../util/named_requirement_verification/legacy
_random_access_iterator.h:14:
In file included from KhronosGroup/SYCL-CTS/tests/accessor/../../util/named_requirement_verification/legacy
_bidirectional_iterator.h:14:
KhronosGroup/SYCL-CTS/tests/accessor/../../util/named_requirement_verification/legacy_forward_iterator.h:11
9:7: error:·
      SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
      WARN(
      ^
KhronosGroup/SYCL-CTS/vendor/Catch2/src/catch2/../catch2/catch_message.hpp:131:23: note:·
      expanded from macro 'WARN'
  #define WARN( msg ) INTERNAL_CATCH_MSG( "WARN", Catch::ResultWas::Warning, Catch::ResultDisposition::ContinueOnFailure, msg )
                      ^
KhronosGroup/SYCL-CTS/vendor/Catch2/src/catch2/../catch2/catch_message.hpp:94:33: note:·
      expanded from macro 'INTERNAL_CATCH_MSG'
        Catch::AssertionHandler catchAssertionHandler( macroName##_catch_sr, CATCH_INTERNAL_LINEINFO, Catch::StringRef(), resul...
                                ^
KhronosGroup/SYCL-CTS/vendor/Catch2/src/catch2/../catch2/internal/catch_assertion_handler.hpp:34:9: note:·
      'AssertionHandler' declared here
        AssertionHandler
        ^
KhronosGroup/SYCL-CTS/tests/accessor/../../util/named_requirement_verification/legacy_forward_iterator.h:51
:3: note:·
      called by 'is_satisfied_for<int *>'
  is_satisfied_for(It valid_iterator, const size_t container_size) {
  ^
```

An example of code triggering the error: https://github.com/KhronosGroup/SYCL-CTS/blob/be8304dcd511707518d0fc7de5ef6a285de2dd4d/util/named_requirement_verification/legacy_forward_iterator.h#L118-L122

Note: we one more similar occurrences in `legacy_bidirectional_iterator.h`.

I haven't prototyped any solution yet, just commented out uses of `WARN` macro in those files simply to proceed further. Before I submit a PR, I would like to get an early feedback to see which solution would be preferable:

1. Drop the warning completely. This could lead to incomplete testing coverage later on
2. Do not support use case when warning is required. Basically, the idea is to fail the test if `container_size` passed to the helper is `0`
3. Convert warning into a message printed through `sycl::stream`
4. Refactor those "requirement" classes so they have separate _host-only_ method for arguments verification: the test must first call that method to (potentially) trigger the warning on host and only then proceed with the test. Well that sounds like (1) with extra steps
5. Drop the `container_size` parameter completely with corresponding branch.

Personally, I would vote for (5), because we anyway don't have any checks in that `container_size == 0` branch and `container_size` doesn't seem to be used anywhere else. Unless I'm missing something or there were further plans to add more checks.

**Comment by AlexeySachkov** [comment]
Moreover, there is a requirement check which explicitly treats `container_size == 0` as an error:

https://github.com/KhronosGroup/SYCL-CTS/blob/be8304dcd511707518d0fc7de5ef6a285de2dd4d/util/named_requirement_verification/legacy_random_access_iterator.h#L180-L184

**Comment by AlexeySachkov** [comment]
I've published #390, which implements approach (5), but I've been thinking that (2) might actually be better in context that it will help prevent adding invalid test cases in the future.
